napari/plugins/tests/test_reader_plugins.py

```diff
@@ -0,0 +1,64 @@
+from napari.plugins.utils import (
+    get_layer_data_from_plugins,
+    permute_hook_implementations,
+)
+from tempfile import NamedTemporaryFile
+import numpy as np
+from skimage import io
+
+
+def test_iter_reader_plugins(plugin_manager):
+    """Test safe iteration through reader plugins even with errors."""
+
+    # first we move one of the "bad" plugins to the front of the line
+    # napari_bad_plugin2 returns a successful "napari_get_reader"
+    # but then raises an IOError
+    permute_hook_implementations(
+        plugin_manager.hook.napari_get_reader, ['napari_bad_plugin2']
+    )
+
+    # the plugin loads fine, so there should be no exceptions yet.
+    assert 'napari_bad_plugin2' not in plugin_manager._exceptions
+
+    # but when we try to read an image path, it will raise an IOError.
+    # we want to catch and store that IOError, and then move on to give other
+    # plugins chance to return layer_data
+    layer_data = get_layer_data_from_plugins('image.ext', plugin_manager)
+
+    # the good plugins (like "napari_test_plugin") should return layer_data
+    assert layer_data
```

Comment:
It might get kind of confusing if we just silently abstract away error messages from the users. They may believe a particular plugin is working when it's not, or not fully understand which plugin did execute. Maybe we could add a Warning when this happens and let the user know that "XYZ" plugin failed to load their data but another was successful?

Reply:
sorry, by "catch and store" I didn't mean to imply it doesn't get reported. A warning _does_ get logged anytime there is a plugin error.  the logic is `plugin.manager. log_plugin_error` and an example of what the warning looks like is here: https://github.com/napari/napari/pull/908#discussion_r380187837
It _also_ gets stored for later retrieval so that they can easily submit a full bug report with traceback to the developer.  I'm working on a full interface for that in another PR.